UI to select a problem location for use in data download

xitij2000 opened the pull request:
This pull request adds a dropdown UI to browse the course structure and select a section, subsection or unit to generate the problem response report for it. 

**Project details**:
This PR is part of the "Improved Problem Response Report" project, funded by Australian National University (ANU) and developed by OpenCraft.

**JIRA tickets**: OSPR-2322 

**Dependencies**: #17776 add support for generating bulk reports. 

**Screenshots**: 
![problem-block-browser](https://user-images.githubusercontent.com/118837/38179272-f0509868-363e-11e8-9107-aadeb771d620.gif)

**Sandbox URL**: 
* LMS: https://pr17824.sandbox.opencraft.hosting/
* Studio: https://studio-pr17824.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:
TBD

**Reviewers**
- [ ] @bradenmacdonald 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [OSPR-2322](https://openedx.atlassian.net/browse/OSPR-2322) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

bradenmacdonald commented on lms/djangoapps/instructor/static/instructor/ProblemBrowser/data/reducers/index.js:
```diff
@@ -0,0 +1,54 @@
+import {combineReducers} from 'redux';
+import {courseBlocksActions} from '../actions/constants';
+
+const BLOCK_TYPES = ['course', 'chapter', 'sequential', 'vertical', 'problem'];
```
People may want to report on types other than `problem`. Ideally in the future, this would be "all structural XBlocks (course, chapter, sequential, vertical), and all XBlocks that implement `generate_report_data()`."

Since we can't do it that way yet, it might be better to do this as a blacklist, excluding `html`, `video`, and `discussion` which are non-reportable, and leaving everything else in by default.

xitij2000 replied:
I've make the bock types that can be filtered by this component configurable via a parameter so it's not hard coded in the component anymore. I feel this makes the component more reusable in places where such a feature might be needed but with different or no filters.

bradenmacdonald commented on lms/djangoapps/instructor/static/instructor/ProblemBrowser/data/store.js:
```diff
@@ -0,0 +1,21 @@
+
+const configureStore = initialState => createStore(
+    rootReducer,
+    initialState,
```
This parameter is meant to be used for "universal apps" that hydrate the state server side and pass in some pre-initialized state to the client side, including some actions already applied.

Here, the reducers each define their own initial state (via the default argument `(state = blah)` in their function definition), so `defaultState` is redundant. You should just remove `defaultState`/`initialState` and let the reducers be the authoritative source of what the default state is.

xitij2000 commented:
@bradenmacdonald  Added screen capture and sandbox links.

xitij2000 commented:
@sstack22 Could you have a review these UI changes?

nedbat commented:
@edx/educator-product Please review.

sstack22 commented:
Will do @xitij2000. CC: @ddomingos - this is an OSPR for the instructor dashboard (improving the problem .csv download report), would you mind taking a look at the UX and letting me know if you think it's good to go?

sstack22 commented:
@bradenmacdonald - I'll sync with the dev team and let you know when we will be ready to look at this.

bradenmacdonald commented:
Great, thanks @sstack22 !

srpearce commented:
@sstack22 and whoever takes this on the engineering team: The UI text needs a doc review (e.g., replace "Pick course block" with "Select a section or problem"; replace "chapter" and "vertical" with "section" and "unit", etc.).

efischer19 commented on lms/djangoapps/instructor/static/instructor/ProblemBrowser/components/BlockBrowser/BlockBrowser.jsx:
```diff
@@ -0,0 +1,58 @@
+/* global gettext */
+import * as React from 'react';
```
🎆 🎉

efischer19 commented on lms/djangoapps/instructor/static/instructor/ProblemBrowser/components/BlockBrowser/BlockBrowser.jsx:
```diff
@@ -0,0 +1,58 @@
+        <span className={classNames('icon fa', `fa-${iconName}`)} aria-hidden="true"/>
+        <span className="sr">{altlabel}</span>
+    </div>
+);
```
Any chance I could convince you to give paragon's `Icon` component a try here?

[Storybook link](https://edx.github.io/paragon/?selectedKind=Icon&selectedStory=with%20screenreader%20text&full=0&addons=1&stories=1&panelRight=1&addonPanel=storybook%2Factions%2Factions-panel)

efischer19 commented on lms/djangoapps/instructor/static/instructor/ProblemBrowser/components/BlockBrowser/BlockBrowser.jsx:
```diff
@@ -0,0 +1,58 @@
+    selectedBlock: PropTypes.string,
+    onSelectBlock: PropTypes.func.isRequired,
+    onChangeRoot: PropTypes.func.isRequired,
+};
```
I'd like to see `BlockList` and `BlockBrowser` defined in a more generic location, do you think they'd be appropriately located in [`common/static/common/js/components`](https://github.com/edx/edx-platform/tree/master/common/static/common/js/components) for consumption by other parts of the platform?

xitij2000 replied:
I've moved these components to the directory you mentioned. I also added it to the module path for web pack to make imports easier. I hope that's OK.

efischer19 commented on lms/djangoapps/instructor/static/instructor/ProblemBrowser/components/Main/Main.jsx:
```diff
@@ -0,0 +1,47 @@
+                    this.hideDropdown();
+                    onSelectBlock(blockId);
+                }}/>}
+            </div>
```
Really enjoying the simplicity of this declaration

efischer19 commented on lms/djangoapps/instructor/static/instructor/ProblemBrowser/data/reducers/index.js:
```diff
@@ -0,0 +1,55 @@
+        }
+        return tree;
+    };
+    return blockTree(blocks.root, null);
```
This is cool 👍

bradenmacdonald commented:
Thanks so much for the review @efischer19 ! Re testing, if there's not support for testing React in the platform yet, I'd recommend as a minimum that we look at adding [Jest Snapshot testing](https://facebook.github.io/jest/docs/en/snapshot-testing.html). It takes almost no work to set up or write tests, and provides a good second check of the HTML output.

xitij2000 commented:
@srpearce  I've updated the strings you pointed out. One concern I do have is that "Select a section or problem" does not capture the fact that even subsections and courses are selectable. Is it worth making this more generic or is it already obvious and I'm thinking too much?

srpearce commented:
@xitij2000 Apologies for the delay—I was out on vacation last week, so I'm just getting to this.

Does this text work?

Button text: **Select Problems**

Select an individual problem to generate a CSV file for that problem, or select a course, section, subsection, or unit to generate a CSV file for all included problems.

Also, @efischer19 : will this option be available by default for all courses on edx.org and Open edX? Will it be live (i.e., visible to course teams) as soon as this PR is merged and released to prod?

xitij2000 commented:
@srpearce  No worries!

I think that test and description makes sense, will update the PR. For the other queryit helps, I can say that this feature is not currently behind a flag.

srpearce commented:
Sorry guys, I just noticed one more thing. Instead of, for example, ``Browsing sequential "Homework - Question Styles":`` when a user has the dropdown list open, I think it would be better to use course/section/subsection/unit—they're much more familiar for most edX course teams.

xitij2000 commented:
@srpearce  Sorry for not updating the screencap yet, but I have updated the text for that with [this recent commit](https://github.com/edx/edx-platform/pull/17824/commits/d502356bdfaaa57f9266faef1daf451815a4449a).

bradenmacdonald commented:
@efischer19 @srpearce @xitij2000 This seems to have stalled - what's the next step to get this merged?

efischer19 commented:
Hey sorry for the delay @bradenmacdonald @xitij2000 - I think we're figured out a plan forward on testing this (and other React components in edx-platform).

I made a barebones Jest setup PR (https://github.com/edx/edx-platform/pull/18364), based on the work you've done here. The plan for now is for us to hash out the details of that work and get it merged onto master separately, then rebase this PR on top of it to add nice tests for this work.

Could you cherry-pick that last commit to enable Jest snapshot testing, then modify the `lms/djangoapps/instructor/static/instructor/ProblemBrowser/components/Main/Main.test.jsx` test and snapshot to be non-trivial?

xitij2000 commented:
Sure, will do!

PS: So sorry for the very late reply! It seems I forgot to actually submit my comment after typing it.

fysheets commented:
Hi @xitij2000 - I took a look at @efischer19's PR mentioned above and made some fixes to get tests passing. If you'd like to cherry pick the commit now with the additional fixes, please do. Or take a look at the PR if you run into test failures from the original commit. Thanks!

xitij2000 commented:
@fysheets Thanks! Sorry for not making much progress here, I hope to come back to the work soon.

xitij2000 commented:
@fysheets  @efischer19 
I've added some tests, and it seems to be working well! Needed to make some changes to the test setup. You'd probably want the setupTests.js file moved elsewhere but I couldn't think of a good place.

fysheets commented:
Sorry for the delay, I completely missed your last message. Thanks for adding the tests! I'll scope out some time to go through a review, meanwhile would you mind checking out these other test failures.

xitij2000 commented:
jenkins run all

xitij2000 commented on package.json:
```diff
@@ -64,12 +64,17 @@
   },
   "devDependencies": {
     "@edx/stylelint-config-edx": "1.1.0",
+    "babel-jest": "^23.0.1",
```
I'm not sure what's causing the issues with the python tests, they seem to be coming from the cherry-picked change from #18364 not really sure how to fix that, some help there would be appreciated.

I've made linting fixes to the code. I also did ran eslint to automatically fix some of the issues it can in the other code. It's in a separate commit so I can remove that if you feel it's a bit much for this PR.

xitij2000 commented:
@fysheets The tests have now passed fully, just rebased to latest master and squashed it down to fewer commits. Hopefully, more tests don't fail now. :crossed_fingers:

xitij2000 commented:
@MichaelRoytman Thanks!

bradenmacdonald commented:
@fysheets @efischer19 @MichaelRoytman Can I go ahead and merge this now? Want us to squash first?

MichaelRoytman commented:
@bradenmacdonald Yes, this is good to go; sorry if I didn't make that clear. If you could squash first, that would be great :).

bradenmacdonald commented:
@MichaelRoytman @xitij2000 has squashed this into three commits, one which adds Jest snapshot testing, one which implements this PR, and one with unrelated quality fixes required to get the tests passing. I'll merge now :)

edx-webhook commented:
@xitij2000 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/17824&entry.1484655318=https://github.com/xitij2000&entry.752974735=2018-03-28+20:58&entry.1917517419=2018-07-17+20:44&entry.2133058324=Yes) so we can improve your experience in the future.

grantgoodmanedX commented:
I have published the doc update:
* http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/student_progress/course_answers.html#report-all-learners-answer-submissions
* https://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/latest/student_progress/course_answers.html#report-all-learners-answer-submissions

bradenmacdonald commented:
Nice, thanks @grantgoodmanedX !

xitij2000 commented:
@grantgoodmanedX Thank you :-)
